### PR TITLE
 Construct Shortest Path and Course Schedule II

### PR DESCRIPTION
**Pull request by nparmar1**

### CODE REVIEW

File: constructShortestPath.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
+/*
+For this approach, we can represent the input data as graph where the cities 
+connecting to each other can be reprsented as nodes
+
+and the roads can be represented as edges. 
+Also by looking at the input, we can represent the graph as undirected and build an adjacency list.
+Thus, for finding the shortest path between two cities, we can find the shorest path the same way we would between nodes in our graph.
+A BFS would be a good approach, since we know that using a BFS, you're guaranteed to find the shorest path, so we can go with that approach. 
+*/
```
Comment (Shreshth3) [inline constructShortestPath.js:9]: The unique part about this problem is that it is NOT exactly a standard BFS. You're not just asked to determine the LENGTH of the shortest path. You are supposed to return the ACTUAL PATH itself. Your explanation should talk about how you plan to do this.

In fact, probably a lot of the sentences you said above can be removed, and you should focus more on this point. In the beginning, you could've just said:
"We can model this as an undirected graph where the cities are nodes and the roads are edges. Then, we could easily find the length of the shortest path with a BFS. But, we need to find the path itself.
<insert explanation of how to find path>"

File: constructShortestPath.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+    while (currentNode !== null) {
+        path.push(currentNode);
+        currentNode = prevNodesMap[currentNode];
+    }
+
+    return path.reverse();
+};
+
+const getShortestPath = (startNode, endNode, nodes) => {
```
Comment (Shreshth3) [inline constructShortestPath.js:49]: Code looks good!

File: constructShortestPath.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        const neighbors = graph[node];
+        for (const neighbor of neighbors) {
+            if (visited.has(neighbor)) continue;
+
+            visited.add(neighbor);
+            queue.enqueue({ node: neighbor, prevNode: node });
+        }
+    }
+
+    return createPath(startNode, endNode, prevNodesMap);
```
Comment (Shreshth3) [inline constructShortestPath.js:75]: P: Why does your algorithm return the SHORTEST path? It seems like you could overwrite values in your `prevNodesMap` and that could cause you to return something that is a true path, but is not the shortest path.

File: courseScheduleII.js [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+    for (let node = 0; node < numNodes; node++) {
+        if (indegreeMap[node] === 0) indegreeZeroNodesQueue.enqueue(node);
+    }
+
+    return indegreeZeroNodesQueue;
+};
+
+const getValidOrderNodes = (graph, indegreeMap, numNodes) => {
+    const indegreeZeroNodesQueue = getIndgreeZeroNodesQueue(indegreeMap, numNodes);
```
Comment (Shreshth3) [inline courseScheduleII.js:70]: P: Why do we care about a node's indegree? What does it represent in the context of this problem? Also, what does it mean (in the context of this problem) if a node has indegree 0?

File: courseScheduleII.js [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        const node = indegreeZeroNodesQueue.dequeue();
+
+        // Process node
+        validOrderNodes.push(node);
+
+        // Get neighbors
+        const neighbors = graph[node];
+        for (const neighbor of neighbors) {
+            indegreeMap[neighbor]--;
+            if (indegreeMap[neighbor] === 0) indegreeZeroNodesQueue.enqueue(neighbor);
```
Comment (Shreshth3) [inline courseScheduleII.js:84]: P: When this condition is true, it means something has now happened for the "neighbor" node. What is that?

In other words, what special property about this neighbor allows us to put it on the queue exactly now, and not earlier?

File: courseScheduleII.js [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        }
+    }
+    return validOrderNodes;
+};
+
+const getOrder = (numCourses, prerequisites) => {
```
Comment (Shreshth3) [inline courseScheduleII.js:90]: Code looks good!

File: courseScheduleII.js [hunk 7]
```diff
@@ -0,0 +1,110 @@
+/*
+For this approach, we can represent the input as a graph where in order to take a course, 
+you have to complete a prereq, which is the same as two nodes conntected by an edge.
```
Comment (Shreshth3) [inline courseScheduleII.js:3]: This is called a "dependency graph". It's probably sufficient to say "We can represent the input as a dependency graph".

Note: dependency graphs are always directed, so there's no need to say that it's directed.

File: courseScheduleII.js [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+ 
+In order to traverse to the next node, you have to start with the one connected to it prior.
```
Comment (Shreshth3) [inline courseScheduleII.js:5]: These two lines are not necessary.

File: courseScheduleII.js [hunk 9]
```diff
@@ -0,0 +1,110 @@
+/*
+For this approach, we can represent the input as a graph where in order to take a course, 
+you have to complete a prereq, which is the same as two nodes conntected by an edge.
+ 
+In order to traverse to the next node, you have to start with the one connected to it prior. 
+As you can see, it's going to be a directed graph, since we can also travel in one direction.
+
+Since we are trying to find a valid ordering for all these courses that respects our prereqs,
```
Comment (Shreshth3) [inline courseScheduleII.js:8]: You really didn't explain anything about HOW this algorithm works. Conceptually, what is it doing?

If you don't know the answer, I suggest that you find a YouTube video and/or work through this code on an example input. That will help you develop a deeper understanding of what this algorithm does. Then, try again at explaining what it's doing.

Here are 4 links I found from just a quick search (I haven't watched these so I don't know if they're good):
https://www.youtube.com/watch?v=rZv_jHZva34
https://www.youtube.com/watch?v=73sneFXuTEg
https://www.youtube.com/watch?v=iI4HhxTZWO8
https://www.youtube.com/watch?v=XMiIk1B0v-w

**Review by Shreshth3 — CHANGES_REQUESTED** [review]

File: constructShortestPath.js (hunk 3 above)
Comment (nparmar1) [reply constructShortestPath.js:75]: It's because when we reach a node for the first time, you are guaranteed that will be the shortest path. Also, the node that previously added the current node, is the ideal node to come from. Since it's the ideal node to come from, there shouldn't be anything overwritten in our object.

File: courseScheduleII.js (hunk 4 above)
Comment (nparmar1) [reply courseScheduleII.js:70]: We care about the node's indegree because it gives us the number of inward directed graph edges associated with that node. For this problem, it represents the prereqs required to take a particular class. For example, class A requires the prereq class B, so class A has an indegree of 1. If a node has an indegree of 0, it means there are no inward edge for that node.

File: courseScheduleII.js (hunk 5 above)
Comment (nparmar1) [reply courseScheduleII.js:84]: When the condition becomes true, it means that they are 0 indegree nodes for that particular node and it can be added to the queue.  Since we are only concerned about the valid order of classes to take, we keep decreasing that class until there's no more prereqs to take. Thus, we will have an ideal class to take in order once the entire iteration is done.

File: constructShortestPath.js (hunk 1 above)
Comment (nparmar1) [reply constructShortestPath.js:10]: Updated

File: courseScheduleII.js [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+we can approach this problem with a topological sort algorithm where if there happens to be an
+edge between a prereq and a class, it means that the prereq appears before the class in a linear ordering fashion.
+
+A class that has a prereq before would be an indegree relationship and we can the can use our topological sort algorithm to track
+the number of indegrees. We can create a hashmap, in this case a JavsScript object, that maps all the classes
+to its prereqs and decrease the number of indegrees once we process that class.
+
+That way we're left with 0 indegree classes in a valid order.
+*/
+
```
Comment (nparmar1) [inline courseScheduleII.js:15]: My latest changes removed your previous comment once I pushed my code. But if you look at the previous commit, you can see your comment. I basically updated my answer for the lines above.

File: constructShortestPath.js [hunk 11]
```diff
@@ -0,0 +1,94 @@
+/*
+We can represent the input data as a graph where the cities are reprsented as nodes and the roads as edges. 
+Also by looking at the input, we can represent the graph as undirected and build an adjacency list.
+Thus, for finding the shortest path between two cities, we can find the shorest path the same way 
+we would between nodes in our graph using a BFS.
+
+But we need to find the path itself. So when we reach a node for the first time, the node previously
+would be the ideal node to come from since it will be the shorest path. Thus, be can build a object
```
Comment (Shreshth3) [inline constructShortestPath.js:8]: This sentence "So when we reach a node for the first time, the node previously would be the ideal node to come from since it will be the shorest path." is really the meat of your answer, and you need to do more to explain WHY this is true. It also should be stated more clearly; this sentence is pretty confusing.

File: constructShortestPath.js (hunk 3 above)
Comment (Shreshth3) [reply constructShortestPath.js:76]: The statements you said are all correct, and they are critical things to understand with this algorithm, so good job. However, you have not given the correct answer to my question about overwriting values. There is a different reason for why the overwriting is not an issue. See if you can figure it out!

File: courseScheduleII.js [hunk 12]
```diff
@@ -0,0 +1,113 @@
+/*
+For this approach, we can represent the input as a dependency graph where in order to take a course,
```
Comment (Shreshth3) [inline courseScheduleII.js:2]: This is not phrased very well. The statement "in order to take a course, you have to complete a prereq" implies that each course has 1 prereq, which is not true. I'd say something like this: "For this approach, we can represent our input as a dependency graph. For each course, all of the incoming edges represent prereqs (i.e. courses we must complete BEFORE taking our current course)."

Try rewriting something like that in your own words.

File: courseScheduleII.js [hunk 13]
```diff
@@ -0,0 +1,113 @@
+/*
+For this approach, we can represent the input as a dependency graph where in order to take a course, 
+you have to complete a prereq, which is the same as two nodes connected by an edge.
+
+Since we are trying to find a valid ordering for all these courses that respects our prereqs,
+we can approach this problem with a topological sort algorithm where if there happens to be an
+edge between a prereq and a class, it means that the prereq appears before the class in a linear ordering fashion.
```
Comment (Shreshth3) [inline courseScheduleII.js:7]: The "where if there happens..." part is not necessary. That is the definition of how a dependency graph works.

File: courseScheduleII.js [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+A class that has a prereq before would be an indegree relationship and we can the can use our topological sort algorithm to track
```
Comment (Shreshth3) [inline courseScheduleII.js:9]: These are implementation details, but you're not describing what this algorithm is doing conceptually.

File: courseScheduleII.js [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+the number of indegrees. We can create a hashmap, in this case a JavsScript object, that maps all the classes
+to its prereqs and decrease the number of indegrees once we process that class.
+
+That way we're left with 0 indegree classes in a valid order.
```
Comment (Shreshth3) [inline courseScheduleII.js:13]: I'm going to do my best to explain how this algorithm works, and hopefully that can improve your intuition.

Let's suppose you were a student at a university, and you want to complete all the required computer science courses at your university. You can model the prerequisite structure as a dependency graph.

Intuitively, how would you approach completing all these courses while meeting the prereqs? Well, you'd do the obvious thing. Find all the courses that have 0 prereqs, and take them during your first semester. Then, during your second semester, find all the classes that either have 0 prereqs, or exclusively have prereqs that you have taken. Take these classes. Next semester, repeat this process.

The idea is that at the start of every semester, you ask "which classes am I now eligible to take" and you take those. In the beginning, that will just be the classes with no prereqs (for example, an introductory programming course). Over time, the set of classes that you're eligible to take will grow larger.

This is pretty much what the BFS topological sort algorithm does (although it is NOT what the DFS topological sort algorithm does). You start by finding nodes with indegree 0. In the context of classes, this is the same as finding all of the classes that have NO prereqs. So, you may notice that indegree is the same thing is as number of prereqs. Then, you "take" all those indegree 0 classes by processing them in your queue. The key detail is that when you "take" a class, you delete all **outgoing** edges from that node. Intuitively, this corresponds to saying "oh, I no longer need to worry about this prerequisite".

Let's consider a concrete example. Suppose we have the following 3 courses:
CS1 - prereqs: none
CS2 - prereqs: CS1
CS3 - prereqs: CS1, CS2

Now you take CS1. Mentally, you can say "oh okay, I no longer need to worry about the prereq of CS1 anywhere." So, you can update your mental model to look like this:
CS2 - prereqs: ~~CS1~~
CS3 - prereqs: ~~CS1~~, CS2
(or you can just delete the CS1 prereqs, which is what our algorithm does).

And you'd repeat this process every time you finish a class.

So, to summarize, your algorithm is just doing the obvious thing that a person would do! Find all the classes with no prereqs and take them. Now, mentally "delete" all the prereqs that you have taken care of, and your procedure. Do this until you've taken all the classes.

I hope this clarified things, and now you can try writing this explanation again. If you're confused, feel free to message me.

File: courseScheduleII.js (hunk 4 above)
Comment (Shreshth3) [reply courseScheduleII.js:73]: You're correct about what indegree represents in the context of this problem. You did not answer what it means if a node has indegree 0 **in the context of this problem**.

File: courseScheduleII.js (hunk 5 above)
Comment (Shreshth3) [reply courseScheduleII.js:87]: Yes, but you should be thinking about the context of this problem.

When this condition is true, it means we have now completed all the prereqs for this class! In others words, we JUST became eligible to take this class.

**Review by Shreshth3 — CHANGES_REQUESTED** [review]

File: constructShortestPath.js (hunk 11 above)
Comment (nparmar1) [reply constructShortestPath.js:8]: So the way we've written our BFS, we only process a node once per node and it will be the first time we've visited that node because of our visited set. And since we're using a BFS to find the actual path, you're guaranteed the shortest path. The current node will always have the ideal node before it which in-turn will give us an actual path from the start node to the end node.

File: constructShortestPath.js (hunk 3 above)
Comment (nparmar1) [reply constructShortestPath.js:76]: The reason why overwriting is not an issue is because of our visited set. The way the algorithm is written is every time we visit a node, it will be the first time and the node previous will always be the ideal node.

File: courseScheduleII.js (hunk 12 above)
Comment (nparmar1) [reply courseScheduleII.js:2]: For this approach, we can represent the input as a dependency graph where each course is followed by one or more prereqs and in order to take to take a course, you have to complete the required courses before the current one. This is the same as node having multiple incoming edges.

File: courseScheduleII.js (hunk 15 above)
Comment (nparmar1) [reply courseScheduleII.js:13]: Thanks! I believe it does. Here's my explanation:

Basically, our algorithm works by first taking all the indegree 0 classes (classes/prereqs that do not have any incoming edges) and adds them to our queue. And for every class that is processed, is it pushed to our valid sorting array. If a particular class has an edge with another class (prereq for a class), we grab that class and decrease the number on indegree by 1 until it's 0 and then added to queue.

File: courseScheduleII.js (hunk 4 above)
Comment (nparmar1) [reply courseScheduleII.js:73]: For the context of this problem, we care about the node being an indegree 0 (when there are no more prereqs to take) because that's when we can push it into our valid sorting array. Thus, gives us a valid topological sort and the ideal ordering of classes to take.

File: courseScheduleII.js (hunk 5 above)
Comment (nparmar1) [reply courseScheduleII.js:87]: Oh, got it! Yes, for the context for this problem, that does make sense.

File: constructShortestPath.js (hunk 11 above)
Comment (Shreshth3) [reply constructShortestPath.js:8]: Sure, this has the right ideas. I'll provide my own description in case this helps.

We've structured our BFS such that we process each node at most one time. We only add a node to the queue the very first time that we see it. Because a BFS goes level-by-level, the first time we see a node, we have used the shortest path to get to that node. Whenever we are adding a neighbor to the queue, we store our current node as the "ideal previous node" for the neighbor. This means that in the shortest path from the start to the neighbor, our current node comes just before the neighbor.

File: constructShortestPath.js (hunk 3 above)
Comment (Shreshth3) [reply constructShortestPath.js:76]: Exactly, we never visit any nodes twice because of the visited set.

File: courseScheduleII.js (hunk 12 above)
Comment (Shreshth3) [reply courseScheduleII.js:2]: Looks good.

File: courseScheduleII.js (hunk 15 above)
Comment (Shreshth3) [reply courseScheduleII.js:13]: Yup, that's right!

File: courseScheduleII.js (hunk 4 above)
Comment (Shreshth3) [reply courseScheduleII.js:73]: Yes, a node with indegree 0 is a class that we are ready to take (we have satisfied the prereqs).

**Review by Shreshth3 — APPROVED** [review]
